pkg/operator/controller/controller.go

```diff
@@ -0,0 +1,377 @@
+
+	// DNSControllerFinalizer is applied to an DNS before being considered for processing;
+	// this ensures the operator has a chance to handle all states.
+	DNSControllerFinalizer = "dns.openshift.io/dns-controller"
+)
+
+// New creates the operator controller from configuration. This is the
+// controller that handles all the logic for implementing dns based on
+// cluster DNS resources.
+//
+// The controller will be pre-configured to watch for DNS resources.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	kubeClient, err := operatorclient.NewClient(config.KubeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kube client: %v", err)
+	}
+
+	reconciler := &reconciler{
+		Config: config,
+		client: kubeClient,
+	}
+	c, err := controller.New("operator-controller", mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &operatorv1.DNS{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	ManifestFactory        *manifests.Factory
+	KubeConfig             *rest.Config
+	CoreDNSImage           string
+	OpenshiftCLIImage      string
+	OperatorReleaseVersion string
+}
+
+// reconciler handles the actual dns reconciliation logic in response to
+// events.
+type reconciler struct {
+	Config
+
+	// client is the kube Client and it will refresh scheme/mapper fields if needed
+	// to detect some resources like ServiceMonitor which could get registered after
+	// the client creation.
+	// Since this controller is running in single threaded mode,
+	// we do not need to synchronize when changing rest scheme/mapper fields.
+	client kclient.Client
+}
+
+// Reconcile expects request to refer to a cluster dns and will do all the work
+// to ensure the cluster dns is in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	errs := []error{}
+	result := reconcile.Result{}
+
+	logrus.Infof("reconciling request: %v", request)
+
+	// Get the current dns state.
+	dns := &operatorv1.DNS{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, dns); err != nil {
+		if errors.IsNotFound(err) {
+			// This means the dns was already deleted/finalized and there are
+			// stale queue entries (or something edge triggering from a related
+			// resource that got deleted async).
+			logrus.Infof("cluster dns not found; reconciliation will be skipped for request: %v", request)
+		} else {
+			errs = append(errs, fmt.Errorf("failed to get cluster dns %s: %v", request, err))
+		}
+		dns = nil
+	}
+
+	if dns != nil {
+		if dns.Name != DefaultDNSController {
+			// We do not want to return this error to avoid re-triggering the event
+			logrus.Errorf("skipping unexpected cluster dns %s", dns.Name)
+			return result, utilerrors.NewAggregate(errs)
```

Comment:
This is technically correct because `errs` must be empty so `utilerrors.NewAggregate(errs)` will return nil, but using `return result, nil` would be more obvious.